calico_rkt/calico_rkt.py

```diff
@@ -14,387 +14,473 @@
+        """
+
+        self.network_name = network_config["name"]
+        """
+        Name of the network from the provided network config file.
+        """
+
+        self.ipam_result = None
+        """
+        Stores the output generated by the IPAM plugin.  This is printed
+        to stdout at the end of execution.
+        """
+
+        # TODO - What config do we need here and how do we get it?
+        # self.calico_config = calico_config
+
+    def parse_cni_args(self, cni_args):
+        """Parses the given CNI_ARGS string into key value pairs
+        and returns a dictionary containing the arguments.
+
+        e.g "FOO=BAR;ABC=123" -> {"FOO": "BAR", "ABC": "123"}
+
+        :param cni_args
+        :return: args_to_return - dictionary of parsed cni args
+        """
+        # Dictionary to return.
+        args_to_return = {}
+
+        # For each arg, split at the equal sign and add to the dictionary.
+        _log.debug("Parsing CNI_ARGS: %s", cni_args)
+        for arg in cni_args.split(";"):
+            _log.debug("\tParsing CNI_ARG: %s", arg)
+            k, v = arg.split("=")
+            args_to_return[k] = v
+        return args_to_return
+
+    def execute(self):
+        """Executes this plugin.
+        Handles unexpected Exceptions in plugin execution.
+
+        :return The plugin return code.
+        """
+        rc = 0
+        try:
+            self._execute()
+        except SystemExit, e:
+            # SystemExit indicates an error that was handled earlier
+            # in the stack.  Just set the return code.
+            rc = e.code 
+        except BaseException:
+            # An unexpected Exception has bubbled up - catch it and
+            # log it out.
+            _log.exception("Unhandled Exception killed plugin")
+            rc = 1
+        finally:
+            _log.debug("Execution complete, rc=%s", rc)
+            return rc
+
+    def _execute(self):
+        """Private method to execute this plugin.
+
+        Uses the given CNI_COMMAND to determine which action to take.
+
+        :return: None.
+        """
+        if self.command == CNI_CMD_ADD:
+            # TODO - If an add fails, we need to clean up any changes we may
+            # have made.
+            self.add()
+        else:
+            assert self.command == CNI_CMD_DELETE, \
+                    "Invalid command: %s" % self.command
+            self.delete()
+
+    def add(self):
+        """"Handles CNI_CMD_ADD requests. 
+
+        Configures Calico networking and prints required json to stdout.
+
+        :return: None.
+        """
+        _log.info('Configuring pod %s' % self.container_id)
+
+        # Step 1: Assign an IP address using the given IPAM plugin.
+        assigned_ip = self._assign_ip()
+
+        # Step 2: Create the Calico endpoint object.
+        endpoint = self._create_endpoint(assigned_ip)
+
+        # Step 3: Provision the veth for this endpoint.
+        endpoint = self._provision_veth(endpoint)
```

Comment:
I think we're missing a step here - see https://github.com/projectcalico/calico-kubernetes/blob/master/calico_kubernetes/calico_kubernetes.py#L342